### ISSUE TITLE
Separate mempool ABCI concerns from mempool data structure/storage engine concerns with clear dependency graph

### ISSUE THREAD

**Issue by jsteenb2** [tendermint/tendermint#9050]
## Summary

As part of an effort to add a new mempool (specifically the narwhal DAG mempool), we are tasked with creating a new mempool. One thing that became immediately obvious as part of this, is that the ABCI concerns are unaffected by the change in storage engine (narwhal mempool vs priority or clist mempool). ATM, all the concerns are bottled up into a single type, which forces the storage engine implementation to do a lot more, increasing its risk of goofing something up in the ABCI layer, which could lead to catastrophic problems. The idea is to separate the concerns better, so that the ABCI concerns are unaffected by the change in storage engine. The ABCI guarantees TM makes, will be enforced regardless the underlying storage engine.

This is part of work we're doing in the PeggyJV/tendermint fork under @zmanian to experiment with high throughput mempools. By increasing the throughput of the mempool we hope to increase the throughput of TM as a whole.

## Problem Definition

We need this feature, b/c we want to make it easier to contribute and experiment with TM. A major area of research across the blockchain community is improving the mempool, as its often the bottle neck. By separating concerns we also create a means to enforce the TM guarantees and reducing risk associated with incorporating a new mempool design.

## Proposal

* Separate ABCI concerns into separate type
* Allow stoage engines to be plug and play without concerning itself with the ABCI layer

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by cmwaters** [comment]
Thanks for opening the issue. I agree with William in regards to formulating an ADR.

I'm curious about what you mean with regards to the storage engine - to me this is the underlying database but the mempool doesn't have any dependency with the database, everything is simply in memory. The mempool is wired up to four components: Consensus, ABCI, P2P & RPC. In prior discussions we've thought about passing the latter 3, connections with the app, the networking layer and the rpc layer either in the constructor or the `Start` method (in a world where we treat things as services). The mempool (whatever implementation it is) chooses to use any of those three components or potentially none at all. The main component mempool is a dependency of is the consensus engine which requires it for providing transactions to be used in blocks. How does this weigh up with your mental model? Would be curious to hear what parts Narwhal uses?

**Comment by jsteenb2** [comment]
agreed, this was primarily created to indicate a need and start a discussion of how we satisfy that need more than anything. I'm currently playing around with implementation details in a fork atm. I'll take a stab at creating a more thorough ADR for this in the near term as well.

**Comment by jsteenb2** [comment]
@cmwaters, I don't have a complete understanding of mempool's involvement in the networking/rpc layer at this moment.  I'll have to dig in more there. I've been primarily focusing on the consensus engine, as that is where this narwhal experiment gets really interesting. With narwhal, we have a "mem"pool that is not in memory. Narwhal is meant to be horizontally scalable, both in terms of storage (narwhal writes to disk) and throughput.  

Narwhal does not speak TXs really, it speaks a language of collection cert digests. The collectiosn themselves being a list of batches. Each batch being a list of TXs. It maintains order as well. In our consensus layer, we won't be sending TXs anymore, we'll be sending little collection cert digests. 

As I was taking this out for a spin, I started with the 0.34.0 release and then after speaking with a few on the team decided to move to the lastest on master. Each branch had a hefty interface for the mempool. It feels like we have a mix of concerns, the ABCI pieces (including the locking for app state and so forth when we execute a commit and the RLock on CheckTx), and then we had the actual data structure or storage engine as I referred to it. As part of narwhal, we don't have a data structure(s), but rather separate processes we speak to over gRPC.  For narwhal, it doesn't impact how ABCI connections should be handled.  Additionally, the caching we do for the app conn in the mempool doesn't need to change. Said another way, the only thing that changes is the data structure/storage engine.

Additionally, we use the reactors to communicate with other mempools in the existing mempool implementaitons (clist/priority), but those are not needed with narwhal. As I think you were hinting at, narwhal maintains its own p2p network for transmitting TXs between nodes.

I'm going to share some videos for narwhal as well.

Does this help shed some light on what the issue is trying to get resolved?
